From: Antonio Terceiro <terceiro@debian.org>
Date: Sat, 27 Aug 2022 23:07:37 -0300
Subject: rbconfig, mkmf: call foreign pkg-config when cross compiling

When cross compiling packages that use pkg-config on Debian, one can use
${DEB_HOST_GNU_TYPE}-pkg-config to find library configuration for the
architecture we are compiling for. For example,
i686-linux-gnu-pkg-config on i386, or aarch64-linux-gnu-pkg-config for
arm64.

For this to work, in Debian we set RUBYLIB to include a directory containing the
foreign rbconfig.rb for the architecture we are building for, e.g.

$ find /usr/lib/aarch64-linux-gnu/ruby-crossbuild -not -type d
/usr/lib/aarch64-linux-gnu/ruby-crossbuild/3.0.0/rbconfig.rb

So when cross compiling, Ruby will load the rbconfig.rb for the
architecture we are compiling for, which is the same mechanism used when
we are cross compiling Ruby itself.

Link: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1018230
Forwarded: https://github.com/ruby/ruby/pull/6355
---
 lib/mkmf.rb      | 2 +-
 tool/mkconfig.rb | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

--- a/lib/mkmf.rb
+++ b/lib/mkmf.rb
@@ -1866,7 +1866,7 @@
     if pkgconfig = with_config("#{pkg}-config") and find_executable0(pkgconfig)
       # if and only if package specific config command is given
     elsif ($PKGCONFIG ||=
-           (pkgconfig = with_config("pkg-config") {config_string("PKG_CONFIG") || "pkg-config"}) &&
+           (pkgconfig = with_config("pkg-config") {config_string("PKG_CONFIG") || (CROSS_COMPILING ? RbConfig::CONFIG["CC"].sub(/-gcc$/, "-pkg-config"): "pkg-config")}) &&
            find_executable0(pkgconfig) && pkgconfig) and
         xsystem([*envs, $PKGCONFIG, "--exists", pkg])
       # default to pkg-config command
--- a/tool/mkconfig.rb
+++ b/tool/mkconfig.rb
@@ -391,7 +391,7 @@
     )
   end
 end
-CROSS_COMPILING = nil unless defined? CROSS_COMPILING
+CROSS_COMPILING = (RbConfig::CONFIG["arch"] != RUBY_PLATFORM) unless defined? CROSS_COMPILING
 EOS
 
 # vi:set sw=2:
